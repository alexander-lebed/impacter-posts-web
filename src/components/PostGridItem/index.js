// @flow
import React, {memo, useState} from 'react';
import LazyLoad from 'react-lazyload';
import ReactImageFallback from 'react-image-fallback';
import isEqual from '../../utils/isEqual';
import changeImageUrlSize from '../../utils/changeImageUrlSize';
import {LOADING_IMAGE} from '../../constants';
import ConfirmationModal from '../common/ConfirmationModal';
import PostModalEdit from '../PostModal/PostModalEdit';
import type {PostType} from '../../types';
import './index.scss';

type Props = {
    post: PostType,
    onClick: Function,
    onEdit: Function,
    onRemove: Function,
}

const Post = (props: Props) => {
    const [showEditModal, setShowEditModal] = useState(false);
    const [showDeleteConfirmation, setDeleteConfirmation] = useState(false);
    const media = props.post.data.media[0];
    const imgUrl = changeImageUrlSize(media.image, 150, 200);
    return (
        <div className="Post">
            <LazyLoad once offset={300} height={150}>
                <ReactImageFallback
                    className="Post-cover"
                    alt="Cover"
                    src={imgUrl}
                    initialImage={LOADING_IMAGE}
                    fallbackImage={LOADING_IMAGE}
                    onClick={() => props.onClick()}
                />
                <div className="Post-text" onClick={() => props.onClick()}>
                    <div className="Post-title">{media.description}</div>
                    <div className="Post-description">{props.post.description}</div>
                </div>
                <div className="Post-actions">
                    <i
                        className="far fa-edit"
                        title="Edit post"
                        onClick={() => setShowEditModal(true)}
                    />
                    <i
                        className="far fa-trash-alt"
                        title="Delete post"
                        onClick={() => setDeleteConfirmation(true)}
                    />
                </div>

                <PostModalEdit
                    isOpen={showEditModal}
                    post={props.post}
                    onEdit={props.onEdit}
                    onBack={() => setShowEditModal(false)}
                    onClose={() => setShowEditModal(false)}
                />

                <ConfirmationModal
                    isOpen={showDeleteConfirmation}
                    onConfirm={() => props.onRemove(props.post.id)}
                    onClose={() => setDeleteConfirmation(false)}
                >
                    <div style={{textAlign: 'center'}}>
                        Delete post "{media.description}"?
                    </div>
                </ConfirmationModal>
            </LazyLoad>
        </div>
    );
};

function areEqual(prevProps, nextProps) {
    return isEqual(prevProps.post, nextProps.post);
}

export default memo(Post, areEqual);