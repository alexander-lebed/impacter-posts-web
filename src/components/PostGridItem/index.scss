.Post {
  display: flex;
  flex-direction: column;
  border: 1px solid black;
  height: 270px;
  width: 200px;
}

.Post-cover {
  cursor: pointer;
  object-fit: cover;
  height: fit-content;
  width: fit-content;
  max-height: 150px;
  max-width: 100%;
}

.Post-text {
  height: 100%;
  cursor: pointer;
  padding: 5px;
  text-align: center;
}

.Post-title {
  font-size: 1.1em;
  font-weight: 500;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.Post-description {
  margin-top: 5px;
  line-height: 1em;
  overflow: hidden;
  text-overflow: ellipsis;
  display: -webkit-box;
  -webkit-box-orient: vertical;
  -webkit-line-clamp: 3;
}

.Post-actions {
  display: flex;
  justify-content: space-between;
  align-items: flex-end;
  flex-grow: 1;
  i {
    cursor: pointer;
    padding: 8px;
  }
}