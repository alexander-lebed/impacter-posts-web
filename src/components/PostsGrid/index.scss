.Posts-loading {
  text-align: right;
}

.Posts-grid {
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(200px, 1fr));
  grid-auto-rows: 280px;
  justify-items: center;
  padding: 5px;
}