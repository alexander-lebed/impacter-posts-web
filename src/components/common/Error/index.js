import React from 'react';

const Error = ({message, className}) => {
    return (
        <div id="error" className={className} style={{color: 'orangered'}}>
            {message}
        </div>
    );
};

export default Error;