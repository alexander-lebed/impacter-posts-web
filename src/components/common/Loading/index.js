import React from 'react';

const Loading = ({className}) => (
    <div id="loading" className={className}>
        Loading...
    </div>
);

export default Loading;