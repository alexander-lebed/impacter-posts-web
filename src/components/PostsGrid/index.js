// @flow
import React, {useState} from 'react';
import Loading from '../common/Loading';
import Error from '../common/Error';
import PostGridItem from '../PostGridItem';
import type {PostType} from '../../types';
import PostModal from '../PostModal';
import {PostDirection} from '../../constants';
import './index.scss';


type Props = {
    loading: boolean,
    error: string,
    posts: {[key: string]: PostType},
    onEdit: (post: PostType) => void,
    onRemove: (id: string) => void
}

const PostsGrid = (props: Props) => {
    const [currentPost, setCurrentPost] = useState<PostType | null>(null);

    const showPost = (direction: $Values<typeof PostDirection>) => {
        if (!currentPost) {
            return;
        }
        const keys = Object.keys(props.posts);
        let targetIndex;
        if (direction === PostDirection.NEXT) {
            targetIndex = keys.indexOf(currentPost.id) + 1;
        } else {
            targetIndex = keys.indexOf(currentPost.id) - 1;
        }
        const postKey = keys[targetIndex];
        if (postKey) {
            setCurrentPost(props.posts[postKey])
        }
    };

    return (
        <div>
            {props.loading && <Loading className='Posts-loading' />}
            {props.error && <Error message={props.error} />}

            <div className="Posts-grid">
                {Object.values(props.posts).map(post => (
                    <PostGridItem
                        key={post.id}
                        post={post}
                        onClick={() => setCurrentPost(post)}
                        onEdit={props.onEdit}
                        onRemove={props.onRemove}
                    />
                ))}
            </div>

            {currentPost && (
                <PostModal
                    post={currentPost}
                    showPost={direction => showPost(direction)}
                    onClose={() => setCurrentPost(null)}
                    onEdit={props.onEdit}
                    onRemove={props.onRemove}
                />
            )}
        </div>
    );
};

export default PostsGrid;