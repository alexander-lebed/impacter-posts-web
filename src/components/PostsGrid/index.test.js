import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme'
import PostsGrid from './index';
import PostGridItem from '../PostGridItem';
import Loading from '../common/Loading';
import Error from '../common/Error';

const dataStr = `{"bcdc3a88-b6cf-4b55-99bd-bf7818032ae2":{"id":"bcdc3a88-b6cf-4b55-99bd-bf7818032ae2","type":"IMAGES","description":"DESCRIPTIOG cillum anim Lorem laborum quis sit aliqua ad occaecat cillum","impacter_id":"822d919d-0076-4d7d-a7ca-c5404a025614","data":{"media":[{"width":5600,"height":3700,"description":"Alejandro","image":"https://picsum.photos/id/0/5516/3700","version":"2020-03-14"}]}},"ce80fa61-89de-4f2e-a2d8-f280bc0573aa":{"id":"ce80fa61-89de-4f2e-a2d8-f280bc0573aa","type":"IMAGES","description":"sint eu nulla fugiat ut in veniam nostrud labore anim","impacter_id":"822d919d-0076-4d7d-a7ca-c5404a025614","data":{"media":[{"width":5616,"height":3744,"description":"Alejandro Escamilla","image":"https://picsum.photos/id/1/5616/3744","version":"2019-03-14"}]}},"0fde87b5-c70c-434a-a6da-90ca149b4e01":{"id":"0fde87b5-c70c-434a-a6da-90ca149b4e01","type":"IMAGES","description":"non in amet id laboris anim et aliquip aliquip ipsum","impacter_id":"822d919d-0076-4d7d-a7ca-c5404a025614","data":{"media":[{"width":2500,"height":1656,"description":"Tina Rataj","image":"https://picsum.photos/id/100/2500/1656","version":"2019-03-14"}]}}}`;


describe('PostGrid', () => {

  const props = {
    posts: JSON.parse(dataStr),
    loading: false,
    error: '',
    onEdit: jest.fn(),
    onRemove: jest.fn(),
  };

  test('should match snapshot', () => {
    const wrapper = renderer.create(<PostsGrid {...props} />);
    const tree = wrapper.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render loading', () => {
    const wrapper = mount(<PostsGrid {...props} loading={true} />);
    const loading = wrapper.find(Loading);
    expect(loading.exists()).toEqual(true);
  });

  it('should render error', () => {
    const wrapper = mount(<PostsGrid {...props} error={'Some error'} />);
    const error = wrapper.find(Error);
    expect(error.exists()).toEqual(true);
    expect(error.text()).toEqual('Some error');
  });

  it('should render posts', () => {
    const wrapper = mount(<PostsGrid {...props} />);
    const postItem = wrapper.find(PostGridItem);
    expect(postItem.exists()).toEqual(true);
    expect(postItem.length).toEqual(3);
  });

});


