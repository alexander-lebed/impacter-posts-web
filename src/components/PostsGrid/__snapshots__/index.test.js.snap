// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostGrid should match snapshot 1`] = `
<div>
  
  <div
    className="Posts-grid"
  >
    <div
      className="Post"
    >
      <div
        className="lazyload-placeholder"
        style={
          Object {
            "height": 150,
          }
        }
      />
    </div>
    <div
      className="Post"
    >
      <div
        className="lazyload-placeholder"
        style={
          Object {
            "height": 150,
          }
        }
      />
    </div>
    <div
      className="Post"
    >
      <div
        className="lazyload-placeholder"
        style={
          Object {
            "height": 150,
          }
        }
      />
    </div>
  </div>
</div>
`;
